plugins/obs-outputs/ftl-stream.c

```diff
@@ -0,0 +1,1188 @@
+
+	while (stream->packets.size) {
+		struct encoder_packet packet;
+		circlebuf_pop_front(&stream->packets, &packet, sizeof(packet));
+
+		/* do not drop audio data or video keyframes */
+		if (packet.type          == OBS_ENCODER_AUDIO ||
+		    packet.drop_priority >= highest_priority) {
+			circlebuf_push_back(&new_buf, &packet, sizeof(packet));
+
+		} else {
+			num_frames_dropped++;
+			obs_encoder_packet_release(&packet);
+		}
+	}
+
+	circlebuf_free(&stream->packets);
+	stream->packets = new_buf;
+
+	if (stream->min_priority < highest_priority)
+		stream->min_priority = highest_priority;
+	if (!num_frames_dropped)
+		return;
+
+	stream->dropped_frames += num_frames_dropped;
+#ifdef _DEBUG
+	debug("Dropped %s, prev packet count: %d, new packet count: %d",
+			name,
+			start_packets,
+			(int)num_buffered_packets(stream));
+#endif
+}
+
+static bool find_first_video_packet(struct ftl_stream *stream,
+		struct encoder_packet *first)
+{
+	size_t count = stream->packets.size / sizeof(*first);
+
+	for (size_t i = 0; i < count; i++) {
+		struct encoder_packet *cur = circlebuf_data(&stream->packets,
+				i * sizeof(*first));
+		if (cur->type == OBS_ENCODER_VIDEO && !cur->keyframe) {
+			*first = *cur;
+			return true;
+		}
+	}
+
+	return false;
+}
+
+static void check_to_drop_frames(struct ftl_stream *stream, bool pframes)
+{
+	struct encoder_packet first;
+	int64_t buffer_duration_usec;
+	size_t num_packets = num_buffered_packets(stream);
+	const char *name = pframes ? "p-frames" : "b-frames";
+	int priority = pframes ?
+		OBS_NAL_PRIORITY_HIGHEST : OBS_NAL_PRIORITY_HIGH;
+	int64_t drop_threshold = pframes ?
+		stream->pframe_drop_threshold_usec :
+		stream->drop_threshold_usec;
+
+	if (num_packets < 5) {
+		if (!pframes)
+			stream->congestion = 0.0f;
+		return;
+	}
+
+	if (!find_first_video_packet(stream, &first))
+		return;
+
+	/* if the amount of time stored in the buffered packets waiting to be
+	 * sent is higher than threshold, drop frames */
+	buffer_duration_usec = stream->last_dts_usec - first.dts_usec;
+
+	if (!pframes) {
+		stream->congestion = (float)buffer_duration_usec /
+			(float)drop_threshold;
+	}
+
+	if (buffer_duration_usec > drop_threshold) {
+		debug("buffer_duration_usec: %" PRId64, buffer_duration_usec);
+		drop_frames(stream, name, priority, pframes);
+	}
+}
+
+static bool add_video_packet(struct ftl_stream *stream,
+		struct encoder_packet *packet)
+{
+	check_to_drop_frames(stream, false);
+	check_to_drop_frames(stream, true);
+
+	/* if currently dropping frames, drop packets until it reaches the
+	 * desired priority */
+	if (packet->priority < stream->min_priority) {
+		stream->dropped_frames++;
+		return false;
+	} else {
+		stream->min_priority = 0;
+	}
+
+	stream->last_dts_usec = packet->dts_usec;
+	return add_packet(stream, packet);
+}
+
+
+static void ftl_stream_data(void *data, struct encoder_packet *packet)
+{
+	struct ftl_stream    *stream = data;
```

Comment:
This struct seems to be meant to be aligned with the declarations below, but it is off by one space when compared to other alignments in this file.